### ISSUE TITLE
JIT solving with shared parameters

### ISSUE THREAD

**Issue by jinfagang** [pytorch/pytorch#22680]
When tracing a model like this:
```
class BasicConv(nn.Module):

    def __init__(self, in_planes, out_planes, kernel_size, stride=1, padding=0, dilation=1,
                 groups=1, relu=True, bn=True, bias=False):
        super(BasicConv, self).__init__()
        self.out_channels = out_planes
        self.conv = nn.Conv2d(in_planes, out_planes, kernel_size=kernel_size,
                              stride=stride, padding=padding, dilation=dilation, groups=groups, bias=bias)
        self.bn = nn.BatchNorm2d(out_planes, eps=1e-5, momentum=0.01, affine=True) if bn else None
        self.relu = nn.ReLU(inplace=True) if relu else None

    def forward(self, x):
        x = self.conv(x)
        if self.bn is not None:
            x = self.bn(x)
        if self.relu is not None:
            x = self.relu(x)
        return x


class FuckNet2(nn.Module):

    def __init__(self):
        super(FuckNet2, self).__init__()

        self.welcome_layer = BasicConv(3, 256, 3, 1, 1)
        self.block_1 = BasicConv(256, 256, 3, 1, 1)
        self.fuck_layers = nn.Sequential()
        for i in range(5):
            self.fuck_layers.add_module('{}'.format(i), self.block_1)

    def forward(self, x):
        x = self.welcome_layer(x)
        return self.fuck_layers(x)
```

I got error:

```
ValueError: TracedModules don't support parameter sharing between modules

```

Why does above net can not trace? In order to trace how should we write model ? I thought my written has no problem........

BTW, this error message has nothing to help, I think it would be better to indicates users where causes the parameters sharing.

**Comment by wanchaol** [comment]
@jinfagang 

Hi, thanks for submitting the issue. That must be a bug on the tracing side, I will investigate. In the meanwhile, you can use our scripting API instead of tracing. If you stay on pytorch nightly or master, you can do the following:

```
with torch.jit._enable_recursive_script(): 
    script_model = torch.jit.script(FuckNet2()) 
```
This will generate the JIT model successfully.

**Comment by wanchaol** [comment]
OK I did some investigation on why this is failing on tracing and this gives me some suspicious 

```
        self.block_1 = BasicConv(256, 256, 3, 1, 1)
        self.fuck_layers = nn.Sequential()
        for i in range(5):
            self.fuck_layers.add_module('{}'.format(i), self.block_1)
```
in here you are appending the `self.block1` to a list multiple times to `self.fuck_layers`, this means you are passing the reference of `self.block1` to the list and in forward function all calls to this list will actually call the same `self.block1`. I don't think this is the correct pattern for your model because most of the `nn.Module`s are stateful, they contain parameters and buffers. In this case you are sharing the parameters and buffers through multiple calls, which is why the tracer gives you this warning. 

Is this something you are expecting to do?

**Comment by jinfagang** [comment]
@wanchaol  Thanks for you reply, this is just an example which can cause this error. In real situation, we always need call same module multi times with a for.. loop such as define a encoder and decoder with multiple scales and multiple levels.

In terms of not using a `self.module` and directly call the `BasicConv`, then how to let them sharing parameters?

The main question here is, I written my model and it trained also inferenced in python, but tracing treated it as error that is not expected. And it claims weights can not sharing between modules but users hard to find which and where layers caused this...


BTW, the nightly can not installed properly:

```
https://download.pytorch.org/whl/nightly/cu100/torch_nightly.html

    return HTMLPage.get_page(link, session=self.session)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 795, in get_page
    resp.raise_for_status()
  File "/usr/share/python-wheels/requests-2.18.4-py2.py3-none-any.whl/requests/models.py", line 935, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://pypi.org/simple/torch-nightly/

```

**Comment by wanchaol** [comment]
Hi @jinfagang sorry for the bad experience, yeah this is a valid pattern in python, but in TorchScript we will need to change the serialization semantics to support such kind of pattern. I think we will fix this in our 1.2 release. 

In order to unblock you, although we could not support putting the same module more than two places in the module hierarchy, you can still call the same module multiple times, that works. 

For the nightly, did you follow the instructions on pytorch.org? It looks like it's a network 404 issue from your error message, I tried it locally on my end it works well.

**Comment by jinfagang** [comment]
@wanchaol  Thank you wanchaol! I just solved my problem yesterday! Turns out to be a module list written format error, now my model runs well!

Appreciate for your help very much!

**Comment by DrStoop** [comment]
> Hi @jinfagang sorry for the bad experience, yeah this is a valid pattern in python, but in TorchScript we will need to change the serialization semantics to support such kind of pattern. **I think we will fix this in our 1.2 release.**

Hi @wanchaol, i just installed the 1.2 release but am still getting the same `ValueError: TracedModules don't support parameter sharing between modules`. Was the fix delayed or is there a timeline?

Thanks & greets!

**Comment by wanchaol** [comment]
> > Hi @jinfagang sorry for the bad experience, yeah this is a valid pattern in python, but in TorchScript we will need to change the serialization semantics to support such kind of pattern. **I think we will fix this in our 1.2 release.**
> 
> Hi @wanchaol, i just installed the 1.2 release but am still getting the same `ValueError: TracedModules don't support parameter sharing between modules`. Was the fix delayed or is there a timeline?
> 
> Thanks & greets!

@DrStoop sorry we made it in a bit later after the 1.2 release, if you check the torch nightly or build from master, this should already works.
